pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/SegmentMergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,200 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentMergeRollupTaskGenerator.class);
+
+  private static final int DEFAULT_MAX_NUM_SEGMENTS_PER_TASK = 20; // 20 segments
+  private static final int DEFAULT_MAX_NUM_RECORDS_PER_SEGMENT = 5_000_000; // 5 million rows
+  private static final int DEFAULT_MAX_NUM_TASKS = 200; // schedule at most 200 tasks at a time
+  private static final String DEFAULT_BUFFER_TIME_PERIOD = "14d"; // 2 weeks
+
+  private final ClusterInfoProvider _clusterInfoProvider;
+
+  public SegmentMergeRollupTaskGenerator(ClusterInfoProvider clusterInfoProvider) {
+    _clusterInfoProvider = clusterInfoProvider;
+  }
+
+  @Override
+  public String getTaskType() {
+    return MinionConstants.MergeRollupTask.TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+
+    // Get the segments that are being converted so that we don't submit them again
+    Map<String, List<String>> scheduledSegmentsMap =
+        TaskGeneratorUtils.getScheduledSegmentsMap(MinionConstants.MergeRollupTask.TASK_TYPE, _clusterInfoProvider);
+
+    for (TableConfig tableConfig : tableConfigs) {
+      // Only generate tasks for OFFLINE tables
+      String offlineTableName = tableConfig.getTableName();
+      if (tableConfig.getTableType() != TableType.OFFLINE) {
+        LOGGER.warn("Skip generating MergeRollupTask for non-OFFLINE table: {}", offlineTableName);
+        continue;
+      }
+
+      TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
+      Preconditions.checkNotNull(tableTaskConfig);
+      Map<String, String> taskConfigs =
+          tableTaskConfig.getConfigsForTaskType(MinionConstants.MergeRollupTask.TASK_TYPE);
+      Preconditions.checkNotNull(taskConfigs, "Task config shouldn't be null for Table: {}", offlineTableName);
+
+      int tableMaxNumTasks =
+          readIntConfigWithDefaultValue(taskConfigs, MinionConstants.TABLE_MAX_NUM_TASKS_KEY, DEFAULT_MAX_NUM_TASKS);
+
+      int maxNumSegmentsPerTask =
+          readIntConfigWithDefaultValue(taskConfigs, MinionConstants.MergeRollupTask.MAX_NUM_SEGMENTS_PER_TASK_KEY,
+              DEFAULT_MAX_NUM_SEGMENTS_PER_TASK);
+
+      int maxNumRecordsPerSegment =
+          readIntConfigWithDefaultValue(taskConfigs, MinionConstants.MergeRollupTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY,
+              DEFAULT_MAX_NUM_RECORDS_PER_SEGMENT);
+
+      String bufferTimePeriodStr =
+          taskConfigs.getOrDefault(MinionConstants.MergeRollupTask.BUFFER_TIME_PERIOD_KEY, DEFAULT_BUFFER_TIME_PERIOD);
+      long bufferTimePeriodInMillis;
+      try {
+        bufferTimePeriodInMillis = TimeUtils.convertPeriodToMillis(bufferTimePeriodStr);
+      } catch (IllegalArgumentException e) {
+        LOGGER.warn(
+            "Buffer time period ('{}') for table '{}' is not configured correctly. Falling back to default behavior ('{}')",
```

Comment:
would it be better to just fail the task in this case

Reply:
+1 to skipping the table (at least initially) if it is misconfigured. A metric can be bumped. I am not sure if we can reverse the effects of the merge in case the table owner comes back with "oops, I meant to do XXX"